File: hello-php/src/Client/Core/WebSocketClient.php
<?php

namespace AudioStreamClient\Core;

use AudioStreamClient\Logger;
use WebSocket\Client;

class WebSocketClient
{
    private string $uri;
    private ?Client $client = null;

    public function __construct(string $uri)
    {
        $this->uri = $uri;
    }

    public function connect(): void
    {
        Logger::info('Connecting to ' . $this->uri);
        $this->client = new Client($this->uri, [
            'fragment_size' => 16384, // 16KB - larger than our 8KB chunks to avoid fragmentation
            'timeout' => 30,
        ]);
        Logger::info('Connected to server');
    }

    public function sendText(array $message): void
    {
        $json = json_encode($message);
        Logger::debug('Sending: ' . $json);
        $this->client->text($json);
    }

    public function sendBinary(string $data): void
    {
        Logger::debug('Sending binary data: ' . strlen($data) . ' bytes');
        // Use send() with 'binary' opcode
        $this->client->send($data, 'binary');
    }

    public function receiveText(): string
    {
        $msg = $this->client->receive();
        Logger::debug('Received: ' . substr($msg, 0, 100));
        return $msg;
    }

    public function receiveBinary(): string
    {
        $msg = $this->client->receive();
        
        // Check if this is actually a text message (JSON error)
        if (is_string($msg) && strlen($msg) > 0 && ($msg[0] === '{' || $msg[0] === '[')) {
            Logger::debug('Received text instead of binary: ' . $msg);
            throw new \Exception('Expected binary data but received text: ' . $msg);
        }
        
        Logger::debug('Received binary data: ' . strlen($msg) . ' bytes');
        return $msg;
    }

    public function close(): void
    {
        if ($this->client !== null) {
            $this->client->close();
            Logger::info('Disconnected from server');
        }
    }
}

File: hello-kotlin/src/main/kotlin/client/core/WebSocketClient.kt
package client.core

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.websocket.*
import io.ktor.websocket.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import WebSocketMessage
import Logger
import kotlin.time.Duration.Companion.seconds

/**
 * WebSocket client for audio streaming
 */
class WebSocketClient(private val uri: String) {
    private val client = HttpClient(CIO) {
        install(WebSockets) {
            pingInterval = 20.seconds
            maxFrameSize = Long.MAX_VALUE
        }
    }
    
    private lateinit var session: DefaultClientWebSocketSession
    
    suspend fun <T> withConnection(block: suspend () -> T): T {
        Logger.info("Connecting to $uri")
        
        return try {
            var result: T? = null
            client.webSocket(uri) {
                session = this
                Logger.info("Connected to server")
                
                // Read and ignore the initial "connected" message from server
                try {
                    val initialFrame = incoming.receive()
                    if (initialFrame is Frame.Text) {
                        val text = initialFrame.readText()
                        Logger.debug("Received initial message: $text")
                    }
                } catch (e: Exception) {
                    Logger.debug("No initial message from server")
                }
                
                result = block()
            }
            result!!
        } catch (e: Exception) {
            Logger.error("WebSocket error: ${e.message}")
            throw e
        } finally {
            client.close()
            Logger.info("Disconnected from server")
        }
    }
    
    suspend fun sendText(message: WebSocketMessage) {
        val json = Json.encodeToString(message)
        Logger.debug("Sending: $json")
        session.send(Frame.Text(json))
    }
    
    suspend fun sendBinary(data: ByteArray) {
        Logger.debug("Sending binary data: ${data.size} bytes")
        session.send(Frame.Binary(true, data))
    }
    
    suspend fun receiveText(): String {
        val frame = session.incoming.receive()
        return when (frame) {
            is Frame.Text -> {
                val text = frame.readText()
                Logger.debug("Received: $text")
                text
            }
            else -> throw Exception("Expected text message, got: ${frame::class.simpleName}")
        }
    }

    suspend fun receiveBinary(): ByteArray {
        val frame = session.incoming.receive()
        return when (frame) {
            is Frame.Binary -> {
                val data = frame.readBytes()
                Logger.debug("Received binary data: ${data.size} bytes")
                data
            }
            is Frame.Text -> {
                val text = frame.readText()
                Logger.debug("Received text message instead of binary: $text")
                // Try to parse as error response
                try {
                    val jsonParser = Json { ignoreUnknownKeys = true }
                    val msg = jsonParser.decodeFromString<WebSocketMessage>(text)
                    if (msg.type == "ERROR") {
                        throw Exception("Server error: ${msg.message}")
                    }
                } catch (e: Exception) {
                    // Ignore parse errors, fall through to default error
                }
                throw Exception("Expected binary message, got text: $text")
            }
            else -> throw Exception("Expected binary message, got: ${frame::class.simpleName}")
        }
    }
}
